mesop/server/server.py

```diff
@@ -261,6 +265,128 @@ def teardown_clear_stale_state_sessions(error=None):
 
   if not prod_mode:
 
+    @flask_app.route("/__editor__/page-commit", methods=["POST"])
+    def page_commit() -> Response:
+      check_editor_access()
+
+      try:
+        data = request.get_json()
+      except json.JSONDecodeError:
+        return Response("Invalid JSON format", status=400)
+      code = data.get("code")
+      path = data.get("path")
+      page_config = runtime().get_page_config(path=path)
+      assert page_config
+      module = inspect.getmodule(page_config.page_fn)
+      assert module
+      module_file = module.__file__
+      assert module_file
+      module_file_path = module.__file__
+      assert module_file_path
+      with open(module_file_path, "w") as file:
+        file.write(code)
+
+      response_data = {"message": "Page commit successful"}
+      return Response(
+        json.dumps(response_data), status=200, mimetype="application/json"
+      )
+
+    @flask_app.route("/__editor__/save-interaction", methods=["POST"])
+    def save_interaction() -> Response:
+      check_editor_access()
+
+      data = request.get_json()
+      if not data:
+        return Response("Invalid JSON data", status=400)
+
+      try:
+        req = urllib_request.Request(
+          "http://localhost:43234/save-interaction",
+          data=json.dumps(data).encode("utf-8"),
+          headers={"Content-Type": "application/json"},
+        )
+        with urllib_request.urlopen(req) as response:
+          if response.status == 200:
+            folder = json.loads(response.read().decode("utf-8"))["folder"]
+            response_data = {"folder": folder}
+            return Response(
+              json.dumps(response_data), status=200, mimetype="application/json"
+            )
+          else:
+            print(f"Error from AI service: {response.read().decode('utf-8')}")
+            return Response(
```

Comment:
[nit] Could be wrong, but think you can return `error_message, status` But I think what's here is a bit more clear.